### PR TITLE
Avoid allocations in Registry.GetBaseKeyFromKeyName

### PR DESCRIPTION
**Pull request by justinvp**
Avoid allocations in `Registry.GetBaseKeyFromKeyName` for improved perf.

<details>
<summary>10,000,000 iteration microbenchmark results:</summary>
```c#
using System;
using System.Diagnostics;
using Microsoft.Win32;

class Program
{
    private static void Main()
    {
        TimeAction("HKEY_USERS");
        TimeAction("HKEY_CLASSES_ROOT");
        TimeAction("HKEY_CURRENT_USER");
        TimeAction("HKEY_LOCAL_MACHINE");
        TimeAction("HKEY_CURRENT_CONFIG");
        TimeAction("HKEY_PERFORMANCE_DATA");
    }

    private static void TimeAction(string keyName)
    {
        string subKeyName;

        Console.WriteLine("Looking for: {0}", keyName);
        TimeAction("Old", () => GetBaseKeyFromKeyNameOld(keyName, out subKeyName));
        TimeAction("New", () => GetBaseKeyFromKeyNameNew(keyName, out subKeyName));
        Console.WriteLine();
    }

    private static void TimeAction(string prefix, Action action, int times = 5, int iterations = 10000000)
    {
        var sw = new Stopwatch();
        for (int i = 0; i < times; i++)
        {
            int gen0 = GC.CollectionCount(0);
            sw.Restart();
            for (int iter = 0; iter < iterations; iter++)
            {
                action();
            }
            sw.Stop();
            Console.WriteLine($"{prefix}: Time: {sw.Elapsed.TotalSeconds}\tGC0: {GC.CollectionCount(0) - gen0}");
        }
    }

    private static RegistryKey GetBaseKeyFromKeyNameOld(string keyName, out string subKeyName)
    {
        if (keyName == null)
        {
            throw new ArgumentNullException(nameof(keyName));
        }

        int i = keyName.IndexOf('\\');
        string basekeyName = i != -1 ?
            keyName.Substring(0, i).ToUpperInvariant() :
            keyName.ToUpperInvariant();

        RegistryKey basekey = null;
        switch (basekeyName)
        {
            case "HKEY_CURRENT_USER":
                basekey = CurrentUser;
                break;
            case "HKEY_LOCAL_MACHINE":
                basekey = LocalMachine;
                break;
            case "HKEY_CLASSES_ROOT":
                basekey = ClassesRoot;
                break;
            case "HKEY_USERS":
                basekey = Users;
                break;
            case "HKEY_PERFORMANCE_DATA":
                basekey = PerformanceData;
                break;
            case "HKEY_CURRENT_CONFIG":
                basekey = CurrentConfig;
                break;
            default:
                throw new ArgumentException(SR.Format(SR.Arg_RegInvalidKeyName, nameof(keyName)), nameof(keyName));
        }

        subKeyName = (i == -1 || i == keyName.Length) ?
            string.Empty :
            keyName.Substring(i + 1, keyName.Length - i - 1);

        return basekey;
    }

    private static RegistryKey GetBaseKeyFromKeyNameNew(string keyName, out string subKeyName)
    {
        if (keyName == null)
        {
            throw new ArgumentNullException(nameof(keyName));
        }

        int i = keyName.IndexOf('\\');
        int length = i != -1 ? i : keyName.Length;

        // Determine the potential base key from the length. Note that
        // HKEY_CLASSES_ROOT and HKEY_CURRENT_USER are the same length,
        // so for that case, switch on a char at a position with a
        // unique char to determine the potential base key.
        RegistryKey baseKey = null;
        switch (length)
        {
            case 10:
                baseKey = Users; // HKEY_USERS
                break;

            case 17:
                const int UniqueCharIndex = 6;
                switch (keyName[UniqueCharIndex])
                {
                    case 'l':
                    case 'L':
                        Debug.Assert(ClassesRoot.Name[UniqueCharIndex] == 'L');
                        baseKey = ClassesRoot; // HKEY_C[L]ASSES_ROOT
                        break;

                    case 'u':
                    case 'U':
                        Debug.Assert(CurrentUser.Name[UniqueCharIndex] == 'U');
                        baseKey = CurrentUser; // HKEY_C[U]RRENT_USER
                        break;
                }
                break;

            case 18:
                baseKey = LocalMachine; // HKEY_LOCAL_MACHINE
                break;

            case 19:
                baseKey = CurrentConfig; // HKEY_CURRENT_CONFIG
                break;

            case 21:
                baseKey = PerformanceData; // HKEY_PERFORMANCE_DATA
                break;
        }

        Debug.Assert(baseKey == null || baseKey.Name.Length == length);

        if (baseKey != null && keyName.StartsWith(baseKey.Name, StringComparison.OrdinalIgnoreCase))
        {
            subKeyName = (i == -1 || i == keyName.Length) ?
                string.Empty :
                keyName.Substring(i + 1, keyName.Length - i - 1);

            return baseKey;
        }

        throw new ArgumentException(SR.Format(SR.Arg_RegInvalidKeyName, nameof(keyName)), nameof(keyName));
    }

    private static readonly RegistryKey CurrentUser = Registry.CurrentUser;
    private static readonly RegistryKey LocalMachine = Registry.LocalMachine;
    private static readonly RegistryKey ClassesRoot = Registry.ClassesRoot;
    private static readonly RegistryKey Users = Registry.Users;
    private static readonly RegistryKey PerformanceData = Registry.PerformanceData;
    private static readonly RegistryKey CurrentConfig = Registry.CurrentConfig;

    private static class SR
    {
        public static readonly string Arg_RegInvalidKeyName = "";
        public static string Format(string format, params object[] args) => "";
    }
}
```
</details>

```
Looking for: HKEY_USERS
Old: Time: 0.8788987    GC0: 114
Old: Time: 0.8631488    GC0: 114
Old: Time: 0.8632139    GC0: 115
Old: Time: 0.8610054    GC0: 114
Old: Time: 0.8777991    GC0: 115
New: Time: 0.1460478    GC0: 0
New: Time: 0.1435679    GC0: 0
New: Time: 0.1483825    GC0: 0
New: Time: 0.1460772    GC0: 0
New: Time: 0.1466389    GC0: 0

Looking for: HKEY_CLASSES_ROOT
Old: Time: 1.0584964    GC0: 152
Old: Time: 1.0440053    GC0: 153
Old: Time: 1.0485327    GC0: 152
Old: Time: 1.0397064    GC0: 153
Old: Time: 1.0480944    GC0: 153
New: Time: 0.1783978    GC0: 0
New: Time: 0.1783887    GC0: 0
New: Time: 0.1766059    GC0: 0
New: Time: 0.1766773    GC0: 0
New: Time: 0.1754843    GC0: 0

Looking for: HKEY_CURRENT_USER
Old: Time: 1.0171277    GC0: 152
Old: Time: 1.0149935    GC0: 153
Old: Time: 0.9913258    GC0: 152
Old: Time: 0.9867099    GC0: 153
Old: Time: 0.9848178    GC0: 153
New: Time: 0.1799622    GC0: 0
New: Time: 0.1788742    GC0: 0
New: Time: 0.1795678    GC0: 0
New: Time: 0.1801555    GC0: 0
New: Time: 0.1844545    GC0: 0

Looking for: HKEY_LOCAL_MACHINE
Old: Time: 1.012618     GC0: 152
Old: Time: 1.010238     GC0: 153
Old: Time: 1.0330128    GC0: 152
Old: Time: 1.0363928    GC0: 153
Old: Time: 0.9934053    GC0: 153
New: Time: 0.1783099    GC0: 0
New: Time: 0.176293     GC0: 0
New: Time: 0.178086     GC0: 0
New: Time: 0.1771917    GC0: 0
New: Time: 0.176571     GC0: 0

Looking for: HKEY_CURRENT_CONFIG
Old: Time: 1.0658348    GC0: 152
Old: Time: 1.0611163    GC0: 153
Old: Time: 1.0657168    GC0: 152
Old: Time: 1.0883544    GC0: 153
Old: Time: 1.0980891    GC0: 152
New: Time: 0.1878427    GC0: 0
New: Time: 0.1887987    GC0: 0
New: Time: 0.1885428    GC0: 0
New: Time: 0.1877254    GC0: 0
New: Time: 0.1892498    GC0: 0

Looking for: HKEY_PERFORMANCE_DATA
Old: Time: 1.1319487    GC0: 172
Old: Time: 1.128515     GC0: 172
Old: Time: 1.1219242    GC0: 171
Old: Time: 1.1288843    GC0: 172
Old: Time: 1.1625333    GC0: 172
New: Time: 0.2075691    GC0: 0
New: Time: 0.2024885    GC0: 0
New: Time: 0.1991121    GC0: 0
New: Time: 0.1958105    GC0: 0
New: Time: 0.1894965    GC0: 0
```

cc: @stephentoub

### CODE REVIEW

File: src/Microsoft.Win32.Registry/tests/Registry/Registry_Getvalue_str_str_obj.cs [hunk 1]
```diff
@@ -12,17 +12,28 @@ namespace Microsoft.Win32.RegistryTests
...
+            new object[] { new string('a', 21) } // String of length 21 that isn't HKEY_PERFORMANCE_DATA.
+        };
 
-            // Should throw if passed string which only starts with one of the valid base key names but actually it isn't valid.
-            Assert.Throws<ArgumentException>(() => Registry.GetValue("HKEY_CURRENT_USER_FOOBAR", null, null));
+        [Theory]
+        [MemberData(nameof(ArgumentExceptionTestData))]
+        public static void InvalidKeyName_ThrowsArgumentException(string keyName)
+        {
+            Assert.Throws<ArgumentException>(() => Registry.GetValue(keyName, null, null));
```
Comment (justinvp) [inline src/Microsoft.Win32.Registry/tests/Registry/Registry_Getvalue_str_str_obj.cs]: Wasn't sure if I should check for the paramName here as desktop doesn't specify it, but CoreFX does.

Comment (stephentoub) [inline src/Microsoft.Win32.Registry/tests/Registry/Registry_Getvalue_str_str_obj.cs]: Best to leave it out in that case.  Thanks.

**Comment by justinvp** [comment]
Test Innerloop CentOS7.1 Debug Build and Test
Test Innerloop CentOS7.1 Release Build and Test

**Comment by justinvp** [comment]
Test Innerloop CentOS7.1 Release Build and Test

**Comment by justinvp** [comment]
Test Innerloop CentOS7.1 Release Build and Test

**Comment by stephentoub** [comment]
Thanks, Justin.  The GC savings is good.  Can you share what kind of effect the throughput increase has on the exposed operations that use this function?  I ask because, while neat and valid, the trick you've applied is certainly more complex, and there are other ways to avoid the allocations, e.g. something like:
```C#
RegistryKey baseKey = null;
if (length > 9) // shortest allowed key
{
    if (string.Compare(keyName, 5, "CURRENT_USER", 0, 12, StringComparison.OrdinalIgnoreCase) == 0) baseKey = CurrentUser;
    else if (string.Compare(keyName, 5, "LOCAL_MACHINE", 0, 13, StringComparison.OrdinalIgnoreCase) == 0) baseKey = LocalMachine;
    ... // etc.
}
```
I've no doubt this will be slower in your microbenchmark than what you've done (I expect it would have similar throughput to the current implementation), but it's also simpler and IMHO easier to reason about, so if the throughput savings isn't pronounced, we might want a simpler approach. Thoughts?

**Comment by justinvp** [comment]
Here's how long it takes to run 100,000 `Registry.GetValue` operations:

```
Looking for value of "Release" in:
HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\NET Framework Setup\NDP\v4\Full

  Old: 0.7895305
  Old: 0.7466162
  Old: 0.7438002
  Old: 0.7383842
  Old: 0.7322767
----------------
New A: 0.7259746
New A: 0.7209606
New A: 0.7222869
New A: 0.7206398
New A: 0.7215663
----------------
New B: 0.7402891
New B: 0.7361964
New B: 0.7343927
New B: 0.7347381
New B: 0.7378687
```

 - "New A" is this PR (slightly tweaked, details below)
 - "New B" is along the lines of what you suggested

Throughput of "New B" is similar to the old implementation as you predicted, whereas "New A" is a small improvement.

### "New B"

```c#
RegistryKey baseKey = null;
if (length > 9) // shortest allowed key
{
    if (string.Compare(keyName, 5, "USERS", 0, 5, StringComparison.OrdinalIgnoreCase) == 0) baseKey = Users;
    else if (string.Compare(keyName, 5, "CLASSES_ROOT", 0, 12, StringComparison.OrdinalIgnoreCase) == 0) baseKey = ClassesRoot;
    else if (string.Compare(keyName, 5, "CURRENT_USER", 0, 12, StringComparison.OrdinalIgnoreCase) == 0) baseKey = CurrentUser;
    else if (string.Compare(keyName, 5, "LOCAL_MACHINE", 0, 13, StringComparison.OrdinalIgnoreCase) == 0) baseKey = LocalMachine;
    else if (string.Compare(keyName, 5, "CURRENT_CONFIG", 0, 14, StringComparison.OrdinalIgnoreCase) == 0) baseKey = CurrentConfig;
    else if (string.Compare(keyName, 5, "PERFORMANCE_DATA", 0, 16, StringComparison.OrdinalIgnoreCase) == 0) baseKey = PerformanceData;
}

if (baseKey?.Name.Length == length && keyName.StartsWith("HKEY_", StringComparison.OrdinalIgnoreCase))
{
    subKeyName = (i == -1 || i == keyName.Length) ?
        string.Empty :
        keyName.Substring(i + 1, keyName.Length - i - 1);

    return baseKey;
}
```

Do you think the approach in this PR would be more palatable if it was simplified a bit and made more compact? e.g.:

### "New A"

```c#
RegistryKey baseKey = null;
switch (length)
{
    case 10: baseKey = Users; break; // HKEY_USERS
    case 17:
        switch (char.ToUpperInvariant(keyName[6]))
        {
            case 'L': baseKey = ClassesRoot; break; // HKEY_C[L]ASSES_ROOT
            case 'U': baseKey = CurrentUser; break; // HKEY_C[U]RRENT_USER
        }
        break;
    case 18: baseKey = LocalMachine; break; // HKEY_LOCAL_MACHINE
    case 19: baseKey = CurrentConfig; break; // HKEY_CURRENT_CONFIG
    case 21: baseKey = PerformanceData; break; // HKEY_PERFORMANCE_DATA
}

if (baseKey != null && keyName.StartsWith(baseKey.Name, StringComparison.OrdinalIgnoreCase))
{
    subKeyName = (i == -1 || i == keyName.Length) ?
        string.Empty :
        keyName.Substring(i + 1, keyName.Length - i - 1);

    return baseKey;
}
```

**Comment by justinvp** [comment]
Or potentially:

```c#
switch (length)
{
    case 10: baseKey = Users; break; // HKEY_USERS
    case 17: baseKey = char.ToUpperInvariant(keyName[6]) == 'L' ? ClassesRoot : CurrentUser; break; // HKEY_C[L]ASSES_ROOT, otherwise HKEY_CURRENT_USER
    case 18: baseKey = LocalMachine; break; // HKEY_LOCAL_MACHINE
    case 19: baseKey = CurrentConfig; break; // HKEY_CURRENT_CONFIG
    case 21: baseKey = PerformanceData; break; // HKEY_PERFORMANCE_DATA
}
```

**Comment by stephentoub** [comment]
> Or potentially:

Ok, let's go with this one. Thanks.

**Comment by justinvp** [comment]
Test Innerloop OSX Debug Build and Test

**Pull request merged**